docs/knowledgebase/advanced/no-hash-collections.md

```diff
@@ -0,0 +1,86 @@
+---
+title: Why are there no Hash collections in sp_std?
+---
+
+Collections such as maps and sets are fundamental data structures. `sp_std` exposes them via
+[`BTreeMap`](https://substrate.dev/rustdocs/v3.0.0/sp_std/collections/btree_map/struct.BTreeMap.html)
+and
+[`BTreeSet`](https://substrate.dev/rustdocs/v3.0.0/sp_std/collections/btree_set/struct.BTreeSet.html)
+respectively. These map closely to the implementations in Rust's standard library. However, the
+standard library also contains
+[`HashMap`](https://doc.rust-lang.org/std/collections/struct.HashMap.html) and
+[`HashSet`](https://doc.rust-lang.org/std/collections/struct.HashSet.html). Let's look into why
+those types are not also exposed in `sp_std`.
+
+## Trees vs Hashes
+
+Trees and hashes are two distinct approaches to implementing the interface of a map or set.
+Fundamentally, trees operate by comparing items: a node in the tree contains an item, and might
+contain references to a set of items less than the node item, and a set of items greater than the
+node item. Tree-based maps and sets have `O(log n)` performance characteristics for `insert`,
+`remove`, and `contains`, where `n` is the number of items currently contained in the collection.
+
+Hashes, on the other hand, lay out a number of "buckets" in memory. Each bucket contains a list of
+items. Their method of operation is to hash each item: reduce it via some hash function into a
+number. Then, they use the hash value to select which bucket should contain the item. Finally, in
+the event that the bucket is not empty, items are handled in unordered manner; the list is scanned
+linearly. Hash-based maps and sets have `O(1)` performance characteristics for `insert`, `remove`,
+and `contains`: their performance does not depend on the number of items they contain.
```

Comment:
you're describing here one strategy of resolving hash collision with separate chaining, 
there is another however, called "open addressing", that doesn't have buckets and in fact rust uses that

https://en.wikipedia.org/wiki/Hash_table#Collision_resolution
https://docs.rs/hashbrown/0.9.1/hashbrown/
https://github.com/abseil/abseil-cpp/blob/master/absl/container/internal/raw_hash_set.h

Reply:
I'd intended to write it generally enough that we can consider the address space of the bucket and its successors as the list which must then be iterated over for that strategy.

The point of this section isn't to teach someone how to implement a hashmap; it's to get them familiar enough that the performance characteristics make sense.